### ISSUE TITLE
IE Driver Server Crashes When Trying to Get Cookies .

### ISSUE THREAD

**Issue by ylupinov** [SeleniumHQ/selenium#8711]
## 🐛 Bug Report

While the specific line of code to get the cookies from the current browser session appears to work, upon trying to get a specific cookie from said collection, the IE Driver Server executable crashes.

## To Reproduce

Our tests get cookies out from browser session to perform some API calls on the background for additional verification. Tests are able to get cookies in Chrome, FF, Safari and also in Safari on IOS, Chrome on Android, but tests are failed on IE

Code:

```
 public static Set<Cookie> getCookies() {
        return Browser.getWebDriver().manage().getCookies();
    }
```

## Expected behavior

IE Driver Server successfully gets the cookies for the current session and a specific cookie can be retrieved.

## Test script or set of commands reproducing this issue
```

 public static Set<Cookie> getCookies() {
        return Browser.getWebDriver().manage().getCookies();
    }
```

Issues without a reproduction script are likely to stall and eventually be closed.

## Environment

OS:
Windows 10 Enterprise (64 bit)
Version 1909
Build: 18363.1016

Selenium Version:
3.141.59

Browser:
Internet Explorer

Browser Version:
11.1016.18362.0

iedriverVersion:
3.141.0
3.150.1

Same issue on the following SauceLabs configuration:
Browser:
iexplore
Browser Version:
11.285.17134.0.
Server OS:
Windows 10
iedriverVersion:
3.141.0
3.150.1

ssue is reproducible running tests on SL, and locally

part of Log.json from SL:
{
"screenshot": null,
"suggestion_values": [],
"start_time": 1600341800.56,
"request": null,
"result": "{here page title goes}",
"duration": 0.031000137329101562,
"path": "title",
"hide_from_ui": false,
"between_commands": 0.06200003623962402,
"visual_command": false,
"HTTPStatus": 200,
"suggestion": null,
"in_video_timeline": 201.55999994277954,
"method": "GET",
"statusCode": 0
},
{
"screenshot": null,
"suggestion_values": [],
"start_time": 1600341800.638,
"request": null,
"result": "https:{here url goes}",
"duration": 0.04699993133544922,
"path": "url",
"hide_from_ui": false,
"between_commands": 0.04699993133544922,
"visual_command": false,
"HTTPStatus": 200,
"suggestion": null,
"in_video_timeline": 201.6380000114441,
"method": "GET",
"statusCode": 0
}
{
"screenshot": null,
"suggestion_values": [],
"start_time": 1600341800.748,
"request": null,
"result": {
"message": "java.net.ConnectException: Connection refused: connect",
"error": "unknown error",
"stackTrace": [
{
"className": "org.openqa.selenium.remote.server.WebDriverServlet",
"lineNumber": 240,
"methodName": "lambda$handle$0",
"fileName": "WebDriverServlet.java"
},
{
"className": "java.util.concurrent.Executors$RunnableAdapter",
"lineNumber": -1,
"methodName": "call",
"fileName": null
},
{
"className": "java.util.concurrent.FutureTask",
"lineNumber": -1,
"methodName": "run",
"fileName": null
},
{
"className": "java.util.concurrent.ThreadPoolExecutor",
"lineNumber": -1,
"methodName": "runWorker",
"fileName": null
},
{
"className": "java.util.concurrent.ThreadPoolExecutor$Worker",
"lineNumber": -1,
"methodName": "run",
"fileName": null
},
{
"className": "java.lang.Thread",
"lineNumber": -1,
"methodName": "run",
"fileName": null
}
]
},
"duration": 34.54200005531311,
"path": "cookie",
"hide_from_ui": false,
"between_commands": 0.06299996376037598,
"visual_command": false,
"HTTPStatus": 500,
"suggestion": null,
"in_video_timeline": 201.74799990653992,
"method": "GET",
"statusCode": 13
}

part from server-log from SL:
11:23:20.748 DEBUG [WebDriverServlet.handle] - Found handler: 7743021c-237f-4bdf-b8c7-45a8f01f756a (org.openqa.selenium.ie.InternetExplorerDriverService)
11:23:20.748 DEBUG [WebDriverServlet.lambda$handle$0] - Handler thread for session 7743021c-237f-4bdf-b8c7-45a8f01f756a (internet explorer): Executing GET on /session/7743021c-237f-4bdf-b8c7-45a8f01f756a/cookie (handler: ServicedSession)
11:23:20.748 DEBUG [HttpURLConnection.writeRequests] - sun.net.www.MessageHeader@4619b118 pairs: {GET /session/7743021c-237f-4bdf-b8c7-45a8f01f756a/cookie HTTP/1.1: null}{Accept: application/json;charset=UTF-8}{Content-Type: application/json;charset=UTF-8}{Cache-Control: no-cache}{Pragma: no-cache}{User-Agent: Java/1.8.0_131}{Host: localhost:33590}{Connection: keep-alive}
11:23:55.352 DEBUG [WebDriverServlet.handle] - Found handler: 7743021c-237f-4bdf-b8c7-45a8f01f756a (org.openqa.selenium.ie.InternetExplorerDriverService)

there was a similar issue before:
#5054

I have tried to downgrade IE driver to version 3.11.1 but have faced the same issue.

**Comment by jimevans** [comment]
@ylupinov In the interest of setting expectations for any taking of action on this issue, without a (publicly available) site to test against that will reproduce the error, it will be nearly impossible to make any changes to the driver to fix it, since the issue as reported does not reproduce for all (or even most) websites. The TRACE-level logs from the “crashed” IEDriverServer.exe session would be the most instructive.

**Comment by ylupinov** [comment]
@jimevans hi, 
i have captured iedriver log for the local run: IEServerlog_local.log
for the sauce labs run it is: selenium-server_ie_cookie_issue.log
[IEServerlog_local.log](https://github.com/SeleniumHQ/selenium/files/5239755/IEServerlog_local.log)
[selenium-server_ie_cookie_issue.log](https://github.com/SeleniumHQ/selenium/files/5239759/selenium-server_ie_cookie_issue.log)


the site to reproduce is https://www.bever.nl
steps are the following:
1) login
2) open lister page (the page with products) https://www.bever.nl/lister.html/ or open product directly, for example, https://www.bever.nl/p/jack-wolfskin-windy-valley-coat-softshell-parka-dames-AEABF00001.html?colour=4012
3) add the product into the basket and click go to the basket on basket pop up
4) in the basket click checkout
5) on the deliver page https://www.bever.nl/buy.html#/delivery perform .`manage().getCookies()`; call

please let me know if you expect to see something else,

**Comment by diemol** [comment]
@ylupinov could you please provide a piece of code using Selenium that goes through those steps? We don't really have the bandwidth to build that code from scratch, and since you already have it, it would be helpful to add it to this issue.

**Comment by ylupinov** [comment]
> @ylupinov could you please provide a piece of code using Selenium that goes through those steps? We don't really have the bandwidth to build that code from scratch, and since you already have it, it would be helpful to add it to this issue.

unfortunately, I can't do that.

**Comment by diemol** [comment]
It does not need to be the same code you are using in your test suite. You can create a script from scratch using Selenium (as you suggested us to do). With that, you won't be exposing any of your internal code.

**Comment by ylupinov** [comment]
> It does not need to be the same code you are using in your test suite. You can create a script from scratch using Selenium (as you suggested us to do). With that, you won't be exposing any of your internal code.

i hope that will help. i don't have much time, but it should work for you to replicate because i was able to do that:

```
public class TestIssue {

    private WebDriver driver;

    @Test 
    public void openSite() throws InterruptedException {
        System.out.println("open site");
        driver.navigate().to("https://www.bever.nl/p/bike-citizens-finn-telefoonhouder-QBAAC90014.html?colour=4168");
        Thread.sleep(15000);
        System.out.println("close coookie pop up");
        driver.findElement(By.xpath("//button[contains(@class,'cookie-message__button')] | //button[@class='as-a-btn as-a-btn--fill']")).click();
        Thread.sleep(10000);
        System.out.println("click account menu");
        driver.findElement(By.xpath("//button[@data-qa='account']")).click();
        Thread.sleep(10000);
        System.out.println("enter credentials");
        driver.findElement(By.xpath("//input[@id='email']")).sendKeys("yiveco1882@gridmauk.com");
        driver.findElement(By.xpath("//input[@id='Password']")).sendKeys("Password!");
        System.out.println("click login");
        driver.findElement(By.xpath("//button[@data-qa='login_btn']")).click();
        Thread.sleep(10000);
        System.out.println("click add to basket");
        driver.findElement(By.xpath("//button[contains(@class,'a-button--add-to-basket')] | //button[@data-qa='add_to_basket']")).click();
        Thread.sleep(10000);
        System.out.println("click go to basket");
        driver.findElement(By.xpath("//a[contains(@class,'button__confirm')] | //a[contains(@href, 'checkout') and not (contains(@data-qa, 'basket'))]")).click();
        Thread.sleep(10000);
        System.out.println("click checkout");
        driver.findElement(By.xpath("//div[contains(@class,'checkout__divider--top')]//button[contains(@class,'button__order')] | //a[@data-qa='checkout_button']")).click();
        Thread.sleep(15000);
        System.out.println("get cookies");
        driver.manage().getCookies();
    }
}
```


execution log:
```
Started InternetExplorerDriver server (32-bit)
3.141.0.0
Listening on port 21434
Log level is set to DEBUG
Log file is set to D:\IEServerlog.log
Only local connections are allowed
Sep 17, 2020 6:34:41 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: W3C


open site
close coookie pop up
click account menu
enter credentials
click login
click add to basket
click go to basket
click checkout
get cookies


org.openqa.selenium.WebDriverException: java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:21434
Build info: version: '3.141.59', revision: 'e82be7d358', time: '2018-11-14T08:17:03'
System info: host: 'EPBYMINW2677', ip: '169.254.97.202', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '1.8.0_251'
Driver info: driver.version: RemoteWebDriver

	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:92)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:552)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:609)
	at org.openqa.selenium.remote.RemoteWebDriver$RemoteWebDriverOptions.getCookies(RemoteWebDriver.java:699)
	at com.asadventure.test.tests.ui.TestIssue.openSite(TestIssue.java:69)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:134)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:597)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:173)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:816)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.testng.TestRunner.privateRun(TestRunner.java:766)
	at org.testng.TestRunner.run(TestRunner.java:587)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:384)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:378)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:337)
	at org.testng.SuiteRunner.run(SuiteRunner.java:286)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1187)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1109)
	at org.testng.TestNG.runSuites(TestNG.java:1039)
	at org.testng.TestNG.run(TestNG.java:1007)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:66)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:109)
Caused by: java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:21434
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:247)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:165)
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:257)
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
	at okhttp3.RealCall.execute(RealCall.java:77)
	at org.openqa.selenium.remote.internal.OkHttpClient.execute(OkHttpClient.java:103)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:155)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:83)
	... 31 more
Caused by: java.net.ConnectException: Connection refused: connect
	at java.net.DualStackPlainSocketImpl.waitForConnect(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:85)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:606)
	at okhttp3.internal.platform.Platform.connectSocket(Platform.java:129)
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:245)
	... 51 more
```
[IEServerlog.log](https://github.com/SeleniumHQ/selenium/files/5240273/IEServerlog.log)

**Comment by diemol** [comment]
Is there no more information at the end of the IEServer log? No errors are shown there 🤔 

```
D 2020-09-17 18:36:19:352 c:\projects\webdriver\cpp\webdriver-server\server.cc(338) Command: GET /session/22bbbbb0-ad26-46d2-8987-802344caf2b6/cookie {}
D 2020-09-17 18:36:19:352 c:\projects\webdriver\cpp\webdriver-server\command.cc(36) Raw JSON command: { "name" : "getAllCookies", "locator" : { "sessionid" : "22bbbbb0-ad26-46d2-8987-802344caf2b6" }, "parameters" : {} }
D 2020-09-17 18:36:19:353 c:\projects\webdriver\cpp\iedriver\iecommandexecutor.cpp(741) Checking for alert before executing getAllCookies command
D 2020-09-17 18:36:19:353 c:\projects\webdriver\cpp\iedriver\iecommandexecutor.cpp(940) No alert handle is found
D 2020-09-17 18:36:19:353 c:\projects\webdriver\cpp\iedriver\iecommandexecutor.cpp(788) Executing command: getAllCookies
I 2020-09-17 18:36:19:353 c:\projects\webdriver\cpp\iedriver\browser.cpp(150) No child frame focus. Focus is on top-level frame
D 2020-09-17 18:36:19:794 c:\projects\webdriver\cpp\iedriver\hookprocessor.cpp(236) Created named pipe \\.\pipe\IEDriverPipe14252
D 2020-09-17 18:36:19:798 c:\projects\webdriver\cpp\iedriver\hookprocessor.cpp(290) Waiting for connection from browser via named pipe
D 2020-09-17 18:36:19:799 c:\projects\webdriver\cpp\iedriver\hookprocessor.cpp(299) Connection from browser established via named pipe
```

**Comment by ylupinov** [comment]
> Is there no more information at the end of the IEServer log? No errors are shown there 🤔
> 
> ```
> D 2020-09-17 18:36:19:352 c:\projects\webdriver\cpp\webdriver-server\server.cc(338) Command: GET /session/22bbbbb0-ad26-46d2-8987-802344caf2b6/cookie {}
> D 2020-09-17 18:36:19:352 c:\projects\webdriver\cpp\webdriver-server\command.cc(36) Raw JSON command: { "name" : "getAllCookies", "locator" : { "sessionid" : "22bbbbb0-ad26-46d2-8987-802344caf2b6" }, "parameters" : {} }
> D 2020-09-17 18:36:19:353 c:\projects\webdriver\cpp\iedriver\iecommandexecutor.cpp(741) Checking for alert before executing getAllCookies command
> D 2020-09-17 18:36:19:353 c:\projects\webdriver\cpp\iedriver\iecommandexecutor.cpp(940) No alert handle is found
> D 2020-09-17 18:36:19:353 c:\projects\webdriver\cpp\iedriver\iecommandexecutor.cpp(788) Executing command: getAllCookies
> I 2020-09-17 18:36:19:353 c:\projects\webdriver\cpp\iedriver\browser.cpp(150) No child frame focus. Focus is on top-level frame
> D 2020-09-17 18:36:19:794 c:\projects\webdriver\cpp\iedriver\hookprocessor.cpp(236) Created named pipe \\.\pipe\IEDriverPipe14252
> D 2020-09-17 18:36:19:798 c:\projects\webdriver\cpp\iedriver\hookprocessor.cpp(290) Waiting for connection from browser via named pipe
> D 2020-09-17 18:36:19:799 c:\projects\webdriver\cpp\iedriver\hookprocessor.cpp(299) Connection from browser established via named pipe
> ```

@diemol unfortunately, that is all we have.because driver crashes.

**Comment by diemol** [comment]
@ylupinov could you please share how the `driver` is created so I can try to reproduce locally?

**Comment by ylupinov** [comment]
> @ylupinov could you please share how the `driver` is created so I can try to reproduce locally?

here it is

```
private DesiredCapabilities ieCapabilities() {
        final DesiredCapabilities capabilities = DesiredCapabilities.internetExplorer();
        capabilities.setCapability(InternetExplorerDriver.ENABLE_ELEMENT_CACHE_CLEANUP, true);
        capabilities.setCapability(InternetExplorerDriver.INTRODUCE_FLAKINESS_BY_IGNORING_SECURITY_DOMAINS, true);
        capabilities.setCapability(InternetExplorerDriver.IGNORE_ZOOM_SETTING, true);
        capabilities.setCapability(InternetExplorerDriver.IE_ENSURE_CLEAN_SESSION, true);
        capabilities.setJavascriptEnabled(true);
        return capabilities;
    }

 private RemoteWebDriver getIEWebDriver(MutableCapabilities desiredCapabilities) {
            System.setProperty(IE_DRIVER_PATH, getIEDriverPath());
            InternetExplorerOptions options = new InternetExplorerOptions();
            options.merge(desiredCapabilities);
            driver = new InternetExplorerDriver(options);
    }

```

**Comment by diemol** [comment]
I tried to reproduce the issue by running the script several times and I was not able to get the same issue. I suspect this might be related to the machine setup. Does it happen in any machine with IE 11? What if you set up a VM from scratch?

This is the [script](https://github.com/diemol/selenium-issues/tree/master/selenium/issue-8711) I used in my attempt to reproduce the issue. I also recorded it to show the output.

![issue_8711](https://user-images.githubusercontent.com/5992658/111343586-dbb36a80-867b-11eb-9897-40e98ebae480.gif)

If you are still having this issue after trying out on different machines or after setting a new VM, and using the latest version of the bindings, feel free to comment and we'll reopen.
